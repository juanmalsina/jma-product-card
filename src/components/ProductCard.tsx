import React from "react";
import { createContext } from "react";
import { useProduct } from "../hooks/useProducts";
import {
  ProductCardProps,
  ProductContextProps,
} from "../interfaces/interfaces";
import styles from "../styles/styles.module.css";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy, maxCount, isMaxCountReached, reset } =
    useProduct({
      onChange,
      product,
      value,
      initialValues,
    });
  return (
    <Provider
      value={{
        counter,
        increaseBy,
        maxCount,
        product,
      }}
    >
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,

          reset,
          increaseBy,
        })}
      </div>
    </Provider>
  );
};