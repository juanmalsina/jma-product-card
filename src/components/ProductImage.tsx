import React from 'react';
import { useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

interface ProductImageProps {
  img?: string;
  className?: string;
  style?: React.CSSProperties;
}

export const ProductImage = ({ img, className, style }: ProductImageProps) => {
  const { product } = useContext(ProductContext);
  const imageToShow = product?.img || img || noImage;
  return (
    <img
      style={style}
      className={`${styles.productImg} ${className}`}
      src={imageToShow}
      alt="coffe mug"
    />
  );
};
