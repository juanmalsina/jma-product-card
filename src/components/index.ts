import { ProducTitle } from './ProducTitle';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductCard } from './ProductCard';

export { ProducTitle, ProductButtons, ProductImage, ProductCard };
