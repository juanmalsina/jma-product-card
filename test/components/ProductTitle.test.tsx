import React from 'react';
import renderer from 'react-test-renderer';
import { ProducTitle, ProductCard } from '../../src/components';
import { product1 } from '../../src/mocks/data';

describe('ProducTitle', () => {
  test('debe mostrar el componente correctamente con el titulo personalizado', () => {
    const wrapper = renderer.create(
      <ProducTitle title="testTitle" className="test-class" />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe de mostrar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard
        initialValues={{ maxCount: 10, count: 0 }}
        product={product1}
      >
        {() => <ProducTitle />}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
