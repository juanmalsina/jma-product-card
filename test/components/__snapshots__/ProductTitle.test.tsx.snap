// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProducTitle Debe de mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    coffee mug - card
  </span>
</div>
`;

exports[`ProducTitle debe mostrar el componente correctamente con el titulo personalizado 1`] = `
<span
  className="productDescription test-class"
>
  testTitle
</span>
`;
